x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/TlsRuntimeCheck.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc.support;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.xpack.core.XPackSettings;
+
+import java.util.function.Consumer;
+
+public class TlsRuntimeCheck {
+
+    private static final Logger logger = LogManager.getLogger(TlsRuntimeCheck.class);
+
+    private final Settings settings;
+    private final boolean httpTlsEnabled;
+    private final boolean transportTlsEnabled;
+
+    public TlsRuntimeCheck(Settings settings) {
+        this.settings = settings;
+        this.httpTlsEnabled = XPackSettings.HTTP_SSL_ENABLED.get(settings);
+        this.transportTlsEnabled = XPackSettings.TRANSPORT_SSL_ENABLED.get(settings);
+    }
+
+    public void checkTlsThenExecute(Consumer<Exception> exceptionConsumer, String featureName, Runnable andThen) {
+        if (false == httpTlsEnabled || false == transportTlsEnabled) {
```

Comment:
This enforces TLS even when bound to localhost, and if bootstrap checks are off.
I think that's too extreme.

Reply:
I was thinking to have the localhost check etc as a separate task, i.e. it will be strict first and relaxed before beta or GA. But we can also tackle it now.

Using the current token service bootstrap check as a reference, I am implementing the logic as follows:
* If `xpack.security.http.ssl.enabled=true`, all good.
* If security is disabled, skip check (this should not happen, but added for completeness)
* If all binding addresses are local, skip check.
* If discovery type is "single-node", skip check.
* Otherwise throw error.